contrib/babelfishpg_tsql/runtime/functions.c

```diff
@@ -3435,3 +3510,415 @@ objectproperty_internal(PG_FUNCTION_ARGS)
+												bbf_pivot_hash,
+												tupdesc,
+												rsinfo->allowedModes & SFRM_Materialize_Random);
+
+	/*
+	 * SFRM_Materialize mode expects us to return a NULL Datum. The actual
+	 * tuples are in our tuplestore and passed back through rsinfo->setResult.
+	 * rsinfo->setDesc is set to the tuple description that we actually used
+	 * to build our tuples with, so the caller can verify we did what it was
+	 * expecting.
+	 */
+	rsinfo->setDesc = tupdesc;
+	MemoryContextSwitchTo(oldcontext);
+
+	oldcontext = MemoryContextSwitchTo(tsql_outmost_context);
+	/* TODO: do we need to free the list */
+	tsql_outmost_estat->pivot_parsetree_list = list_delete_nth_cell(tsql_outmost_estat->pivot_parsetree_list, tsql_outmost_estat->pivot_number - 1);
+	tsql_outmost_estat->pivot_number--;
+	MemoryContextSwitchTo(oldcontext);
+	return (Datum) 0;
+}
+
+/*
+ * load up the categories hash table
+ */
+static HTAB *
+load_categories_hash(RawStmt *cats_sql, MemoryContext per_query_ctx)
+{
+	HTAB	   *bbf_pivot_hash;
+	HASHCTL		ctl;
+	int			ret;
+	uint64		proc;
+	MemoryContext SPIcontext;
+
+	/* initialize the category hash table */
+	ctl.keysize = MAX_CATNAME_LEN;
+	ctl.entrysize = sizeof(bbf_pivot_HashEnt);
+	ctl.hcxt = per_query_ctx;
+
+	/*
+	 * use INIT_CATS, defined above as a guess of how many hash table entries
+	 * to create, initially
+	 */
+	bbf_pivot_hash = hash_create("bbf_pivot hash",
+								INIT_CATS,
+								&ctl,
+								HASH_ELEM | HASH_STRINGS | HASH_CONTEXT);
+
+	/* Connect to SPI manager */
+	if ((ret = SPI_connect()) < 0)
+		/* internal error */
+		elog(ERROR, "load_categories_hash: SPI_connect returned %d", ret);
+
+	/* Retrieve the category name rows */
+	ret = SPI_execute_raw_parsetree(cats_sql, true, 0);
+	proc = SPI_processed;
+
+	/* Check for qualifying tuples */
+	if ((ret == SPI_OK_SELECT) && (proc > 0))
+	{
+		SPITupleTable *spi_tuptable = SPI_tuptable;
+		TupleDesc	spi_tupdesc = spi_tuptable->tupdesc;
+		uint64		i;
+
+		/*
+		 * The provided categories SQL query must always return one column:
+		 * category - the label or identifier for each column
+		 */
+		if (spi_tupdesc->natts != 1)
+			ereport(ERROR,
+					(errcode(ERRCODE_SYNTAX_ERROR),
+					 errmsg("provided \"categories\" SQL must " \
+							"return 1 column of at least one row")));
+
+		for (i = 0; i < proc; i++)
+		{
+			bbf_pivot_cat_desc *catdesc;
+			char	   *catname;
+			HeapTuple	spi_tuple;
+
+			/* get the next sql result tuple */
+			spi_tuple = spi_tuptable->vals[i];
+
+			/* get the category from the current sql result tuple */
+			catname = SPI_getvalue(spi_tuple, spi_tupdesc, 1);
+			if (catname == NULL)
+				ereport(ERROR,
+						(errcode(ERRCODE_SYNTAX_ERROR),
+						 errmsg("provided \"categories\" SQL must " \
+								"not return NULL values")));
+			
+			SPIcontext = MemoryContextSwitchTo(per_query_ctx);
+			catdesc = (bbf_pivot_cat_desc *) palloc(sizeof(bbf_pivot_cat_desc));
+			catdesc->catname = catname;
```

Comment:
You need to do a deep copy of `catname` because you are doing a `SPI_finish()` before returning `bbf_pivot_hash`. You can try to confirm this by running one of your tests that go to this codepath with valgrind enabled.

Reply:
This part was originally from tablefunc.c’s load_categories_hash, and it is a good finding. The reason why it does not causing any problem is we only use catname as a argument of snprintf to generate a key for the hash table. snprintf(key, MAX_CATNAME_LEN - 1, "%s", CATNAME); \And later on the retrieval from bbf_pivot_HashTableLookup, we got catdesc, and we only used the catdesc->attidx but not the catdesc->catname. catname at this point is pointing to an random address but we are not using it, so we are not breaking anything, but it is still a very good finding.